CMakeLists.txt

```diff
@@ -131,10 +131,21 @@ include("cmake/host_compiler.cmake")
 include("cmake/configuring_primitive_list.cmake")
 
 if(UNIX OR MINGW)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99")
-    # Let SYCL to choose the C++ standard it needs.
-    if(NOT DNNL_WITH_SYCL)
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+    if(CMAKE_VERSION VERSION_LESS "3.1.0")
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99")
+        # Let SYCL to choose the C++ standard it needs.
+        if(NOT DNNL_WITH_SYCL)
+            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+        endif()
+    else()
+        # Use CMake 3.1 functionality and allow overriding to
+        # higher values
+        if(NOT CMAKE_C_STANDARD OR CMAKE_C_STANDARD_DEFAULT EQUAL 89)
+            set(CMAKE_C_STANDARD 99)
+        endif()
+        if(NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD_DEFAULT EQUAL 98)
```

Comment:
I think it will break when `DNNL_WITH_SYCL` is `TRUE`. The `CMAKE_<LANG>_STANDARD` variables don't seem to have a default value hence they are empty. In the case of SYCL it will lead to using C++11, which is not what we want.

Reply:
SYCL should set `CMAKE_CXX_STANDARD` to the value it wants.

That's what I did for my project. Adding `DNNL_WITH_<projectname>` for a dozen projects does not scale.

Reply:
SYCL is not a project, it's just a runtime that oneDNN uses when a users requests it. The runtime is supported via Intel DPC++ compiler that doesn't set any CMake variables.
The rule of thumb is to pass `-fsycl` option to the DPC++ compiler and let the compiler decide on the C++ standard. That's what we do. This change breaks build for SYCL runtime when CMake 3.1 or newer is used.

Reply:
That's not an issue, then. If the compiler chooses the right C++ standard when passed `-fsycl`, then `CMAKE_CXX_STANDARD_DEFAULT` will be the correct value.

Reply:
It is actually an issue because the default standard for DPC++ compiler is C++14 however when you specify `-fsycl` option the compiler automatically enables C++17 if another standard is not explicitly specified. If you do `set(CMAKE_CXX_STANDARD 11)` then the DPC++ compiler will use C++11.

Reply:
Understood, but this section won't trigger because with that compiler, `CMAKE_CXX_STANDARD_DEFAULT` will be 14, not 98.

Wait, the "OR" should be "AND"